samples:Host Exerciser Enhancement

ramyavex opened the pull request:
Added fpgaperf_counter library and fpgaperf class to host exerciser source code to report similar metrics as fpgadiag does in OFS EA.

Implemented fpgaperf_counter shared library module works similar to perf tool.

It reads the FPGA performance counter values before workload and after workload, calculate delta and prints.

[fpgaperf_counter_log.txt](https://github.com/OPAE/opae-sdk/files/6429268/fpgaperf_counter_log.txt)
[he_enhancement_log.txt](https://github.com/OPAE/opae-sdk/files/6429269/he_enhancement_log.txt)

tswhison reviewed (commented):
Couple of high-level things need to be taken care of before a more in-depth review.  Please make sure to add the Intel copyright notice to the top of all files, including CMakeLists.txt.  Use existing files as examples.  Secondly, the code formatting is mostly incorrect according to our standards.  Please use checkpatch.pl from the Linux kernel tree and fix all errors that are reported.  Thirdly, since you are linking against opae-c, you can take advantage of the OPAE_INFO, OPAE_ERR, and OPAE_DBG printing macros.  No need to re-invent your own in the header.  Thanks!

ramyavex commented:
Updated the review comments

    Added header to CMakelist.txt
    Formatted the source using checkpatch.pl
    Used OPAE_ERR instead of PERF_ERR.

    Working on unit test cases.

anandaravuri commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,461 @@
+{
+	DIR *dir = NULL, *event_dir = NULL, *format_dir = NULL;
+	struct dirent *entry = NULL;
+	char dir_name[MAX] = "";
```
Initialize like  char path[SYSFS_PATH_MAX] = { 0 };  dir_name ,event_name

ramyavex commented:
Added the unit test cases for fpgaperf_counter library.
Attached the output logs.

[output_for_hw_test_log.txt](https://github.com/OPAE/opae-sdk/files/6460932/output_for_hw_test_log.txt)
![output_of_unit_test_fpgaperf](https://user-images.githubusercontent.com/77089807/117858624-b0c84a00-b2ab-11eb-9e06-a9db583a2ff8.PNG)

tswhison commented on samples/host_exerciser/host_exerciser_cmd.h:
```diff
@@ -26,12 +26,55 @@
+  }
+  fpga_result perfstop(void)
+  {
+	fpga_result res = FPGA_OK;
+
+	res = fpgaperfcounterstop();
+	if (res != FPGA_OK)
+		return FPGA_EXCEPTION;
+
+	return FPGA_OK;
```
```c
return fpgaperfcounterstop();
```

ramyavex replied:
fixed

ramyavex commented:
Fixed the review comments

    Added mutex lock and unlock mechanism for global structure
    For print function added FILE * parameter
    Modified OPAE_ERR format
    Modified the strtol to return last byte instead of character
    Modified the return statements

ramyavex commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,508 @@
+};
+
+/* Not static so other tools can access the PMU data */
+struct fpga_perf_counter *fpga_perf_pmus = NULL;
```
Added opae mutex lock and unlock mechanism for global structure

ramyavex commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,508 @@
+	for (loop = 0; loop < (int)rdft->nr; loop++) {
+		for (inner_loop = 0; inner_loop < fpga_perf_pmus->num_generic_events; inner_loop++) {
+			if (rdft->values[loop].id == fpga_perf_pmus->generic_events[inner_loop].id) {
+				if (is_start == 1)
```
Modified

ramyavex commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,508 @@
+	res = read_fab_counters(fpga_perf_pmus->is_start);
+	if (res != FPGA_OK)
+		return FPGA_EXCEPTION;
+	return FPGA_OK;
```
fixed

ramyavex commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,508 @@
+	return FPGA_OK;
+}
+
+void fpgaperfcounterprint(void)
```
added FILE * parameter for fpgaperfcounterprintf()

ramyavex commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,508 @@
+	FILE *file = NULL;
+	char sysfspath[DFL_BUFSIZ_MAX] = { 0,};
+
+	if ((dir == NULL) || (dir_name == NULL)){
```
fixed

ramyavex commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,508 @@
+	/* reset the performance counter to 0 */
+	if (ioctl(fpga_perf_pmus->generic_events[0].fd, PERF_EVENT_IOC_RESET, PERF_IOC_FLAG_GROUP) == -1) {
+		OPAE_ERR("PERF_EVENT_IOC_RESET ioctl failed: %s", strerror(errno));
+		return FPGA_EXCEPTION;
```
fixed

ramyavex commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,508 @@
+
+void fpgaperfcounterfree(void)
+{
+	if (fpga_perf_pmus->formats != NULL)
+		free(fpga_perf_pmus->formats);
+	if (fpga_perf_pmus->generic_events != NULL)
+		free(fpga_perf_pmus->generic_events);
+	if (fpga_perf_pmus != NULL)
+		free(fpga_perf_pmus);
```
fixed

ramyavex commented on samples/host_exerciser/host_exerciser_cmd.h:
```diff
@@ -26,12 +26,55 @@
+	if (res != FPGA_OK)
+		return FPGA_EXCEPTION;
+
+	return FPGA_OK;
```
fixed

ramyavex commented on samples/host_exerciser/host_exerciser_cmd.h:
```diff
@@ -26,12 +26,55 @@
+	if (res != FPGA_OK)
+		return FPGA_EXCEPTION;
+
+	return FPGA_OK;
```
fixed

tswhison reviewed (commented):
Use of global variable fpga_perf_pmus is proving difficult with respect to locking.  Instead of making this variable a global, why not make the first parameter to each function here a pointer to a struct fpga_perf_counter?  This would eliminate the need to perform locking around the global, and is much cleaner.

tswhison commented on samples/fpgaperf_counter/fpgaperf_counter.c:
This is still dereferencing a NULL pointer.  If someone calls..
```c
fpgaPerfCounterPrint(NULL);
```

tswhison commented on samples/fpgaperf_counter/fpgaperf_counter.h:
```diff
@@ -0,0 +1,93 @@
+ * @returns FPGA_OK on success. FPGA_EXCEPTION if an internal exception occurred
+ * while trying to access the handle.
+ */
+fpga_result fpgaPerfCounterStartrecord(void);
```
fpgaPerfCounterStartRecord

tswhison commented on samples/fpgaperf_counter/fpgaperf_counter.h:
```diff
@@ -0,0 +1,93 @@
+ * @returns FPGA_OK on success. FPGA_EXCEPTION if an internal exception occurred
+ * while trying to access the handle.
+ */
+fpga_result fpgaPerfCounterStoprecord(void);
```
fpgaPerfCounterStopRecord

tswhison commented on tests/fpgaperf/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+        fpgaperf-static
+)
+
+target_link_libraries(test_fpgaperf_c "-ludev")
```
Remove this line and place udev below line 30.

tswhison commented on samples/fpgaperf_counter/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+	COMPONENT opaesamplelib
+)
+
+target_link_libraries(fpgaperf_counter "-ludev")
```
Remove this line and place udev below line 31.

tswhison commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,639 @@
+
+#define DFL_PERF_STR_MAX	256
+
+#define PERF_EVNET_PATTERN	"event=(0x[0-9a-fA-F]{2}),evtype=(0x[0-9a-fA-F]{2}),portid=(0x[0-9a-fA-F]{2})"
```
PERF_EVENT_PATTERN

ramyavex replied:
fixed

tswhison commented on afu-test/afu_test.h:
```diff
@@ -189,6 +189,7 @@ class afu {
     }
 
     auto tokens = fpga::token::enumerate({filter});
+    perf_token_ = tokens[0];
```
Hi Ramya, this change is causing a segfault in test_dummy_afu.  Could you please implement a fix?  Thanks!

ramyavex replied:
getting a token from handler. Its fixed testing error

r-rojo commented on samples/fpgaperf_counter/fpgaperf_counter.c:
Should log a message and return FPGA_INVALID_PARAM

ramyavex replied:
changed as per the comment

r-rojo commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,639 @@
+#include <opae/utils.h>
+#include <opae/fpga.h>
+#include <regex.h>
+#include "fpgaperf_counter.h"
```
Should probably undefine _GNU_SOURCE after including linux headers and also reorder the includes by:
1. prototype file (fpgaperf_counter.h)
2. C system headers (stdio, stdlib, ...)
3. Other standard libraries (libudev.h, linux/perf_event.h, ...)
4. standard opae headers (<opae/...> that are part of OPAE installation
5. other project headers 

Each header group should have its headers listed in alphabetical order with a space between each group.

ramyavex replied:
fixed

r-rojo commented on samples/fpgaperf_counter/fpgaperf_counter.c:
I'm not sure it it matters or not but I would release resources in the opposite order that they were acquired, especially if one resource depends on another acquired before it.

ramyavex replied:
changed the order of releasing resources

r-rojo commented on samples/fpgaperf_counter/fpgaperf_counter.c:
Is this the number at the end? I'm not sure it's in hex.

ramyavex replied:
changed to decimal value

r-rojo commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,639 @@
+
+/* Not static so other tools can access the PMU data */
+fpga_perf_counter *g_fpga_perf = NULL;
+
```
If other tools need to access PMU data then maybe we should provide API functions to get them.
Also, are we assuming we can only have one of these at a time? What if I have multiple cards and I have an application what wants to monitor perf data in multiple threads?

r-rojo commented on samples/fpgaperf_counter/fpgaperf_counter.c:
Does this function assume that  the device this token represents is on the same bdf as the fme? or does it assume that this token is a token for an FME?

tswhison commented on samples/host_exerciser/host_exerciser.h:
```diff
@@ -454,6 +457,12 @@ class host_exerciser : public test_afu {
+  
+  token::ptr_t get_parent_token()
+  {
+    auto tokens = token::enumerate({properties::get(handle_)});
```
How is this getting the parent token?

tswhison commented on samples/host_exerciser/host_exerciser_cmd.h:
```diff
@@ -26,12 +26,40 @@
+  }
+  fpga_result perfprint(fpga_perf_counter *fpga_perf)
+  {
+      FILE *file = stdout;
+
+      return fpgaPerfCounterPrint(file, fpga_perf);
```
```c
return fpgaPerfCounterPrint(stdout, fpga_perf);
```

tswhison commented on tests/fpgaperf/fpgaperf_counter.h:
```diff
@@ -0,0 +1,153 @@
+// Copyright(c) 2021, Intel Corporation
```
This header seems to be duplicated.  Instead of doing that, we should add the necessary include directory when building the tests.

tswhison commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -0,0 +1,646 @@
+	} while (0)
+
+/* mutex to protect fpga perf pmu data structures */
+pthread_mutex_t fpga_perf_lock = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
```
Since we are no longer tracking a global variable, but are instead passing an fpga_perf_counter * to the API functions, we no longer need this global mutex (fpga_perf_lock).  Instead, let's add a pthread_mutex_t to struct fpga_perf_counter and lock/unlock it as needed.

r-rojo commented on samples/fpgaperf_counter/fpgaperf_counter.h:
```diff
@@ -61,6 +70,28 @@ typedef struct {
 } fpga_perf_counter;
 
+/**
+ * Initialize the pthread mutex.
```
I think this is an implementation detail that doesn't really summarize what this function does.
IOW, it's too specific .

r-rojo replied:
Or maybe this really needs to be an internal function (not an API function) that is called during the initialization of the fpga_perf_counter structure.

r-rojo commented on samples/fpgaperf_counter/fpgaperf_counter.h:
```diff
@@ -61,6 +70,28 @@ typedef struct {
+fpga_result fpgaPerfCreateHandle(fpga_perf_counter *fpga_perf);
+
+/**
+ * Destroy the pthread mutex.
```
I think this is an implementation detail that doesn't really summarize what this function does.
IOW, it's too specific .

r-rojo replied:
Also, I'm a little confused why we have two API functions that initialize part of the data structure.

r-rojo commented on tests/fpgaperf/test_fpgaperf_counter_c.cpp:
```diff
@@ -0,0 +1,192 @@
+				t = nullptr;
+			}
+		}
+		delete fpga_perf;
```
If the assertion on line 88 fails then line 89 won't execute (so we shouldn't delete)

r-rojo commented on samples/fpgaperf_counter/fpgaperf_counter.c:
I this this could be a memory leak if we call parse_perf_event multiple times using the same fpga_perf_counter structure.

r-rojo reviewed (commented):
I think we should evaluate the lifecycle of the data structures
(how/when structures are allocated and initialized) and where the responsibility lies (user vs API library).

r-rojo commented on samples/host_exerciser/host_exerciser_cmd.h:
```diff
@@ -117,11 +151,41 @@ class host_exerciser_cmd : public test_command
+                fpgaperf.perfdestroy(fpga_perf);
+                fpgaperf.destroyhandle(fpga_perf);
+                delete fpga_perf;
+                fpga_perf = nullptr;
```
I'm seeing a pattern here when we want to abort and have to cleanup the fpga_perf structure.
Perhaps this could be refactored/wrapped by a function or a c++ object that does lifecycle management?

r-rojo reviewed (commented):
I now understand that the class fpga_perf is a sort of wrapper around the fpga perf APIs.
This class is used without any member variables and its functions behave mostly like static functions. IOW, the functions don't operate on the calling object and only operate on the in/out parameters. Also, the way your code is structured, every time you call a function from this class, you check the return code and if it is non-zero, you have to destroy the C structure before aborting.
I think maybe another way to wrap the perf API in a C++ object that helps you out could look like the snippet below. 
 

```c++
class fpgaperf
{
public:

static std::shared_ptr<fpgaperf> get(token::ptr_t token)
{
  auto p = std::make_shared<fpgaperf>();
  if (fpgaPerfCounterGet(token->c_type(), &p->counter_)){
    // error message or something
    p.reset();
  }
  return p;
}

~fpgaperf()
{
  fpgaPerfCounterDestroy(&counter_);
  // maybe log message if result is non-zero
}

fpga_result start()
{
  return fpgaPerfCounterStartRecord(&counter_);
}
fpga_result stop()
{
  return fpgaPerfCounterStopRecord(&counter_);
}

fpga_result print()
{
  return fpgaPerfCounterPrint(stdout, &counter_);
}

private:
  fpgaperf();
  fpga_perf_counter counter_;
};
```

The nice thing about doing it this way is that the C structure is destroyed when its C++ wrapper instance goes out of scope. In this example I've made the default constructor private so the only way to get the C++ instance is by calling the static method, get.
So the calling function would look something like this:

```c++

auto perf = fpgaperf::get();
if (!perf) {
  // raise error
}

// use the perf object
perf->start();
// stuff
perf->stop();

```

_note_: if you do it this way, you might also want to make the copy constructor private...

ramyavex commented on samples/fpgaperf_counter/fpgaperf_counter.c:
```diff
@@ -450,31 +455,41 @@ fpga_result fpgaPerfCreateHandle(fpga_perf_counter *fpga_perf)
 
-fpga_result fpgaPerfDestroyHandle(fpga_perf_counter *fpga_perf)
+/* Reset the magic number and destroy the mutex created */
+fpga_result fpga_perf_mutex_destroy(fpga_perf_counter *fpga_perf)
```
We can't make because, we are calling this one in unit test.

tswhison replied:
You can use STATIC (in all caps).  For the test build, STATIC is compiled out; but in non-test builds STATIC will evaluate to static.

ramyavex replied:
Updated with STATIC

tswhison reviewed (dismissed):
👍

r-rojo commented on samples/host_exerciser/host_exerciser_cmd.h:
```diff
@@ -36,26 +36,48 @@ namespace host_exerciser {
+        p.reset();
+        return NULL;
+    }
+    return p;
```
I think this is overcomplicating things a bit.
The example I gave was mostly for reference and so I hadn't tested it.
If you're getting compiler errors with make_shared, then I suggest using something simpler:
```c++
  std::shared_ptr<fpgaperf> p(new fpgaperf());
  if (fpgaPerfCounterGet(token->c_type(), p->counter_) != FPGA_OK) {
    p.reset();
  }
  return p;

```

r-rojo commented on samples/host_exerciser/host_exerciser_cmd.h:
```diff
@@ -36,26 +36,48 @@ namespace host_exerciser {
+      return fpgaPerfCounterStopRecord(counter_);
   }
-  fpga_result perfdestroy(fpga_perf_counter *fpga_perf)
+  fpga_result print()
```
When moving to c++, try to simplify the code.
For example, print seems like it should return a void (unless you thing there's value in capturing the error code).
If you happen to catch errors while performing some lower level operation, then you can print the message here.
Also, you might want to switch to using c++ exceptions for error notification and handling instead of error codes.
